pkgs/test/incrementalBuild/default.nix

```diff
@@ -0,0 +1,57 @@
+{ hello, buildIncremental, runCommandNoCC, texinfo, stdenv, rsync }:
+let
+  baseHelloArtifacts = buildIncremental.prepareIncrementalBuild hello;
+  patchedHello = hello.overrideAttrs (old: {
+    buildInputs = [ texinfo ];
+    src = runCommandNoCC "patch-hello-src" { } ''
+      mkdir -p $out
+      cd $out
+      tar xf ${hello.src} --strip-components=1
+      patch -p1 < ${./hello.patch}
+    '';
+  });
+  incrementalBuiltHello = buildIncremental.mkIncrementalBuild patchedHello baseHelloArtifacts;
+
+  incrementalBuiltHelloWithCheck = incrementalBuiltHello.overrideAttrs (old: {
+    doCheck = true;
+    checkPhase = ''
+      echo "checking if unchanged source file is not recompiled"
+        [ "$(stat --format="%Y" lib/exitfail.o)" = "$(stat --format="%Y" ${baseHelloArtifacts}/outputs/lib/exitfail.o)" ]
+    '';
+  });
+
+  baseHelloRemoveFileArtifacts = buildIncremental.prepareIncrementalBuild (hello.overrideAttrs (old: {
+    patches = [ ./hello-additionalFile.patch ];
+  }));
+
+  preparedHelloRemoveFileSrc = runCommandNoCC "patch-hello-src" { } ''
+    mkdir -p $out
+    cd $out
+    tar xf ${hello.src} --strip-components=1
+    patch -p1 < ${./hello-additionalFile.patch}
+  '';
+
+  patchedHelloRemoveFile = hello.overrideAttrs (old: {
+    buildInputs = [ texinfo ];
+    src = runCommandNoCC "patch-hello-src" { } ''
+      mkdir -p $out
+      cd $out
+      ${rsync}/bin/rsync -cutU --chown=$USER:$USER --chmod=+w -r ${preparedHelloRemoveFileSrc}/* .
+      patch -p1 < ${./hello-removeFile.patch}
+    '';
+  });
+
+  incrementalBuiltHelloWithRemovedFile = buildIncremental.mkIncrementalBuild patchedHelloRemoveFile baseHelloRemoveFileArtifacts;
+in
+stdenv.mkDerivation {
+  name = "patched-hello-returns-correct-output";
```

Comment:
I think the tests should be more extensive. This is a very generic function that can be applied to essentially any derivation. Does it work with `chromium`? What about Python packages or Haskell packages? Does it work with all derivations perhaps? Does it work with derivations that mess around with phases? Etc.

Reply:
RE: Haskell packages, you can compare and contrast this PR with my analogous Haskell-specific PR in https://github.com/NixOS/nixpkgs/pull/204020

The one main thing that would not work if you applied this to Haskell builds is that macOS builds would not correctly work.  The reason why is because of the linker work-around on macOS (and the Haskell PR disables the linker workaround when incremental builds are enabled)

I think other than I think that the two approaches could be unified.  The other primary difference between the two is that the Haskell approach only exports the `dist/build` directory and not the entire directory, for efficiency reasons (since there is no benefit in reusing other build products)

Reply:
> The other primary difference between the two is that the Haskell approach only exports the dist/build directory and not the entire directory, for efficiency reasons (since there is no benefit in reusing other build products)

The export of the entire directory is essential in my use case, as some projects like the Linux Kernel  does not have a build directory, where all the build outputs are stored.

But yes, if there is a chance for unification, we should try it.

Reply:
> I think the tests should be more extensive. This is a very generic function that can be applied to essentially any derivation. Does it work with `chromium`? What about Python packages or Haskell packages? Does it work with all derivations perhaps? Does it work with derivations that mess around with phases? Etc.

I understand, why more extensive tests are fine.

In terms of build time, larger tests, such as a Linux kernel build will increase the testing time massively.

I can introduce such tests as well, but i would argue that large tests that tests the same behavior like the small ones are not that beneficial.

In terms of python, or haskell, I have not the knowledge about the build systems that i can argue it works for them as well.